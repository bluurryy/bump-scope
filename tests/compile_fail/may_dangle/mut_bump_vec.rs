//! When we add the `#[may_dangle]` attribute via the "nightly-dropck-eyepatch" feature
//! we need to make sure `T` is still marked as owned and thus as being dropped by the container.
//! We do this by having a `PhantomData<T>` field.
//! If we didn't do this, then this would compile and cause UB because were
//! accessing a dangling reference (use after free).

use core::fmt::Debug;

use bump_scope::{Bump, MutBumpVec};

struct PrintOnDrop<T: Debug>(T);

impl<T: Debug> Drop for PrintOnDrop<T> {
    fn drop(&mut self) {
        std::println!("dropping: {:?}", self.0);
    }
}

fn dangling_reference(bump: &mut Bump) {
    let mut v = MutBumpVec::new_in(bump);
    let s = String::from("hello");
    v.push(PrintOnDrop(&s));
}

fn main() {}
