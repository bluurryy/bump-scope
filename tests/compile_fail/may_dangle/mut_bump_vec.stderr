error[E0597]: `s` does not live long enough
  --> tests/compile_fail/may_dangle/mut_bump_vec.rs:22:24
   |
21 |     let s = String::from("hello");
   |         - binding `s` declared here
22 |     v.push(PrintOnDrop(&s));
   |                        ^^ borrowed value does not live long enough
23 | }
   | -
   | |
   | `s` dropped here while still borrowed
   | borrow might be used here, when `v` is dropped and runs the destructor for type `MutBumpVec<PrintOnDrop<&String>, &mut Bump>`
   |
   = note: values in a scope are dropped in the opposite order they are defined
